[Alerting] Adds navigation by consumer and alert type to alerting (#58997)
elastic/kibana: x-pack/plugins/alerting/README.md

});
```

## Alert Navigation

> When registering an Alert Type, you'll likely want to provide a way of viewing alerts of that type within your own plugin, or perhaps you want to provide a view for all alerts created fro mwithin your solution within your own UI.

Reviewer: ```suggestion
When registering an Alert Type, you'll likely want to provide a way of viewing alerts of that type within your own plugin, or perhaps you want to provide a view for all alerts created from within your solution within your own UI.
```

> In order for the Alerting framework to know that your plugin has its own interna; view for displaying an alert, you must resigter a navigation handler within the framework.

Reviewer: ```suggestion
In order for the Alerting framework to know that your plugin has its own internal view for displaying an alert, you must resigter a navigation handler within the framework.
```

A navigation handler is nothing more than a function that receives an Alert and its corresponding AlertType, and is expected to then return the path *within your plugin* which knows how to display this alert.

The signature of such a handler is:

```

> type AlertNavigationHandler = (

Reviewer 2: the function result type seems to be `JsonObject | string` in `alerting/public/alert_navigation_registry/types.ts`.  Are we not supporting `JsonObject` for now, at this level of the API?  Guessing we can support it, would need more tests and doc, etc, deferred to later ...

Author: This is for the State feature that doesn't work because we're booted by the management plugin which is legacy.
Once this is migrated this feature will work and only then can we add a functional test that would actually pass, which is why I don't mention it in the docs.

But once it's migrated navigation will work out of the box, without us having to do any more code for it - so it seemed worth keeping in.

### registerNavigation
The _registerNavigation_ api allows you to register a handler for a specific alert type within your solution:

```
alerting.registerNavigation(
	'my-application-id',
	'my-application-id.my-alert-type',
	(alert: SanitizedAlert, alertType: AlertType) => `/my-unique-alert/${alert.id}`
);
```

This tells the Alerting framework that, given an alert of the AlertType whose ID is `my-application-id.my-unique-alert-type`, if that Alert's `consumer` value (which is set when the alert is created by your plugin) is your application (whose id is `my-application-id`), then it will navigate to your application using the path `/my-unique-alert/${the id of the alert}`.

The navigation is handled using the `navigateToApp` api, meaning that the path will be automatically picked up by your `react-router-dom` **Route** component, so all you have top do is configure a Route that handles the path `/my-unique-alert/:id`.

You can look at the `alerting-example` plugin to see an example of using this API, which is enabled using the `--run-examples` flag when you run `yarn start`.

> ### registerNavigation

Reviewer 2: s/registerNavigation/registerDefaultNavigation/

Author: Good catch, thanks

The _registerDefaultNavigation_ api allows you to register a handler for any alert type within your solution:

```
alerting.registerDefaultNavigation(
	'my-application-id',
	(alert: SanitizedAlert, alertType: AlertType) => `/my-other-alerts/${alert.id}`
);
```

This tells the Alerting framework that, given any alert whose `consumer` value is your application, as long as  then it will navigate to your application using the path `/my-other-alerts/${the id of the alert}`.

### balancing both APIs side by side

> As we mentioned, using `registerDefaultNavigation` will tell the Alerting Framework that your application can handle any type of Alert we throw at it, as long as your application created it, using the handler your provide it.

Reviewer: ```suggestion
As we mentioned, using `registerDefaultNavigation` will tell the Alerting Framework that your application can handle any type of Alert we throw at it, as long as your application created it, using the handler you provide it.
```
